### PR TITLE
Bug fix/replication fast lane

### PR DESCRIPTION
**Pull request by mchacki**
# Pull Request Guidelines

Pull requests are an essential collaborative tool for modern software development. 

The below list is intended to help you figure out whether your code is ready to be reviewed
and merged into ArangoDB. The overarching goal is to:

- Reduce the amount of *recurring* defects
- Reduce the impact to the other developer’s time and energy spent on defects in your code
- Increase the overall autonomy and productivity of individual developers

## Acceptance Checklist

*The below list is not exhaustive, think thoroughly whether the provided information is sufficient.*
*Remove options that do not apply*

### Scope & Purpose

*This PR should move the replication to fast execution lane. This way a replication insertion will be handled with higher priority as a client requested insertion.*

- [x] Bug-Fix for *devel-branch* (i.e. no need for backports?)
- [x] Bug-Fix for a *3.4* port will happen after we validated in devel that this is ok
- [ ] Strictly *new functionality* (i.e. a new feature / new option, no need for porting)
- [ ] The behavior in this PR can be (and was) *manually tested* (support / qa / customers can test it)
- [x] The behaviour change can only be verified via automatic tests

#### Related Information

*(Please reference tickets / specification etc )*

- [ ] There is a GitHub Issue reported by a Community User: 
- [ ] There is an internal planning ticket: 
- [ ] There is a *JIRA Ticket number* (In case a customer was affected / involved): 
- [ ] There is a design document: 

### Testing & Verification

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This PR adds tests that were used to verify all changes:

- [ ] Added **Regression Tests** (Only for bug-fixes) 
- [x] Added new C++ **Unit Tests** (Either GoogleTest or Catch-Test)
   - Did you add tests for a new *RestHandler* subclass ? Yes
   - Did you add new mocks of underlying code layers to be able to verify your functionality ? No

- [ ] Added new **integration tests** (i.e. in shell_server / shell_server_aql)
- [ ] Added new **resilience tests** (only if the feature is impacted by failovers)

Additionally:

- [x] There are tests in an external testing repository (i.e. node-resilience tests, chaos tests) 
  http://jenkins01.arangodb.biz:8080/job/short-time-perfmance-test/ (currently blocked, will add a specific link tomorrow.
- [ ] I ensured this code runs with ASan / TSan or other static verification tools

*(Include link to Jenkins run etc)*
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/5011/

> Think about whether the new code you added is modular enough to be
> easily testable by unit tests written with GTest / Catch. It is not good if your feature is so interconnected
> that it prevents other people from writing their own unit gests. It should be possible
> to use your code in future without extensively mocking your classes.
> A bad example that required some extensive effort would be the storage engine API.

### Documentation

> All new Features should be accompanied by corresponding documentation. 
> Bugs and features should furthermore be documented in the changelog so that
> developers and users have a concise overview. 

- [x] Added a *Changelog Entry*
(Nothing is changed form user-perspective)
- [ ] Added entry to *Release Notes* 
- [ ] Added a new section in the *Manual* 
- [ ] Added a new section in the *http API* 
- [ ] Added *Swagger examples* for the http API  

### CLA Note 

Please note that for legal reasons we require you to sign the [Contributor Agreement](https://www.arangodb.com/documents/cla.pdf)
before we can accept your pull requests.

### CODE REVIEW

File: tests/Mocks/Servers.cpp [hunk 1]
```diff
@@ -159,9 +168,45 @@ std::unique_ptr<arangodb::aql::Query> MockAqlServer::createFakeQuery() const {
...
-      false, getSystemDatabase(), fakeQueryString, bindParams, queryOptions,
-      arangodb::aql::QueryPart::PART_DEPENDENT);
+  auto query =
+      std::make_unique<arangodb::aql::Query>(false, getSystemDatabase(),
+                                             fakeQueryString, bindParams, queryOptions,
+                                             arangodb::aql::QueryPart::PART_DEPENDENT);
   query->injectTransaction(createFakeTransaction());
   return query;
 }
+
```
Comment (joerg84) [inline tests/Mocks/Servers.cpp]: Is this here related to your change?

File: tests/Mocks/Servers.cpp [hunk 2]
```diff
@@ -25,28 +25,30 @@
...
 #include "RestServer/QueryRegistryFeature.h"
 #include "RestServer/SystemDatabaseFeature.h"
 #include "RestServer/TraverserEngineRegistryFeature.h"
 #include "RestServer/ViewTypesFeature.h"
-#include "StorageEngine/EngineSelectorFeature.h"
 #include "Sharding/ShardingFeature.h"
+#include "StorageEngine/EngineSelectorFeature.h"
 #include "Transaction/Methods.h"
 #include "Transaction/StandaloneContext.h"
+#include "V8Server/V8DealerFeature.h"
```
[the remark this answers is not in the document]
Comment (mchacki) [reply tests/Mocks/Servers.cpp]: My editor does alphabetic sorting of them as we decided and as based on clang format.
I am happy to include this in the CI pipline or as a commit hook, however it seems we have some subtle differences in clang-format versions (using same config) so we need to nail down one specific version first

File: arangod/RestHandler/RestDocumentHandler.h [hunk 3]
```diff
@@ -35,7 +35,16 @@ class RestDocumentHandler : public RestVocbaseBaseHandler {
  public:
   RestStatus execute() override final;
   char const* name() const override final { return "RestDocumentHandler"; }
-  RequestLane lane() const override final { return RequestLane::CLIENT_SLOW; }
+  RequestLane lane() const override final {
+    bool found = false;
```
[the remark this answers is not in the document]
Comment (mchacki) [reply arangod/RestHandler/RestDocumentHandler.h]: :+1:

File: tests/Mocks/Servers.cpp [hunk 4]
```diff
@@ -159,9 +168,45 @@ std::unique_ptr<arangodb::aql::Query> MockAqlServer::createFakeQuery() const {
...
+  startFeatures();
+}
+
+MockRestServer::~MockRestServer() {
+  arangodb::LogTopic::setLogLevel(arangodb::iresearch::TOPIC.name(),
+                                  arangodb::LogLevel::DEFAULT);
+  arangodb::LogTopic::setLogLevel(arangodb::Logger::FIXME.name(), arangodb::LogLevel::DEFAULT);
+  arangodb::LogTopic::setLogLevel(arangodb::Logger::AUTHENTICATION.name(),
+                                  arangodb::LogLevel::DEFAULT);
+}
```
Comment (maierlars) [inline tests/Mocks/Servers.cpp]: missing new line at EOF! 😱

**Comment by joerg84** [comment]
@fceller Could you make sure we cover this with performance regression tests?

File: CHANGELOG [hunk 5]
```diff
@@ -1,6 +1,9 @@
 devel
 -----
 
+* Replication requests on Document API are now on higher priority then client-triggered requests.
+  This should help to keep in sync repliction up and running even if the server is overloaded.
```
Comment (jsteemann) [inline CHANGELOG]:
```suggestion
  This should help to keep in sync replication up and running even if the server is overloaded.
```

File: arangod/RestHandler/RestDocumentHandler.h [hunk 6]
```diff
@@ -35,7 +35,16 @@ class RestDocumentHandler : public RestVocbaseBaseHandler {
  public:
   RestStatus execute() override final;
   char const* name() const override final { return "RestDocumentHandler"; }
-  RequestLane lane() const override final { return RequestLane::CLIENT_SLOW; }
+  RequestLane lane() const override final {
+    bool isSyncReplication = false;
+    // We do not care for the real value, enough if it is there.
+    _request->value(StaticStrings::IsSynchronousReplicationString, isSyncReplication);
```
Comment (goedderz) [inline arangod/RestHandler/RestDocumentHandler.h]: You could emphasize the comment:
```suggestion
    std::ignore = _request->value(StaticStrings::IsSynchronousReplicationString, isSyncReplication);
```
